man/coap_encryption.txt.in

```diff
@@ -84,20 +84,306 @@ In principle the set-up sequence for CoAP Clients looks like
+these reserved definitions.
+
+*SECTION: PSK Client: coap_dtls_cpsk_t: Identity Hint Callback*
+[source, c]
+----
+/**
+ * Identity Hint Validation call-back that can be set up by
+ * coap_new_client_session_psk2().
+ * Invoked when libcoap has done the validation checks at the TLS level,
+ * but the application needs to check that the Identity Hint is allowed,
+ * and needs to use the appropriate PSK key for the (D)TLS session.
+ * Note: This is not supported in (D)TLS1.3.
+ *
+ * @param hint  The server provided Identity Hint (should be zero terminated)
+ * @param hint_len  The length of the Identity Hint (not including zero)
+ * @param coap_session  The CoAP session associated with the Identity Hint
+ * @param arg  The same as was passed into coap_new_client_session_psk2()
+ *             in setup_data->ih_call_back_arg
+ *
+ * @return New coap_dtls_cpsk_key_t object or @c NULL on error.
+ */
+typedef coap coap_dtls_cpsk_key_t *(*coap_dtls_ih_callback_t)(
+                                const uint8_t *hint,
+                                size_t hint_len,
+                                coap_session_t *coap_session,
+                                void *arg);
+----
+
+*validate_ih_call_back* points to an application provided Identity Hint callback
+function or NULL. The application can make use of this Identity Hint information
+to decide what Identity and PSK should be used for this session.
+The Callback returns the new coap_dtls_cpsk_key_t on success,
+or NULL if the Identity Hint is unacceptable.
+
+*NOTE:* The Server may provide NULL hint, or a zero length hint to indicate
+there is no hint.  In this case the initially provided Identity and PSK should
+be used.
+
+*ih_call_back_arg* points to a user defined set of data that will get passed
+in to the validate_ih_call_back() function's arg parameter and can be used by
+that function.  An example would be a set of Identity Hints that map into new
+Identity / PSKs to use.
+
+*SECTION: PSK Client: coap_dtls_cpsk_t: Subject Name Indicator (SNI) Definition*
+
+*client_sni* points to the SNI name that will be added in as a TLS extension,
+or set NULL.  This typically is the DNS name of the server that the client is
+trying to contact.  The server is then able to decide, based on the name in the
+SNI extension, whether, for example, a different Hint and/or PSK is to be used.
+
+Note: Not supported by TinyDTLS.
+
+*SECTION: PSK Client: coap_dtls_cpsk_t: PSK Key Definitions*
+[source, c]
+----
+typedef struct coap_dtls_cpsk_key_t {
+  const uint8_t *identity;   /* The Identity (should be zero terminated */
+  unsigned int identity_len; /* The Identity length (not including the zero) */
+  const uint8_t *key;        /* The PSK to use */
+  unsigned int key_len;      /* The PSK length */
+} coap_dtls_cpsk_key_t;
+----
+
+*identity* points to the data to use for the identity. This cannot be NULL.
+
+*identity_len* is the length of the identity. This cannot be 0.
```

Comment:
Hm, according to [RFC 4279](https://tools.ietf.org/html/rfc4279#section-2), the identity may have zero length.

Reply:
Corrected and updated coap-client to support `u=""` for testing.